src/proposal/mod.rs

```diff
@@ -0,0 +1,151 @@
+pub mod repost;
+pub mod timeline;
+
+use std::collections::HashSet;
+
+use self::timeline::TimelineStatus;
+
+use crate::notify::get_text_mentions;
+use crate::str_serializers::*;
+use crate::{Balance, SponsorshipToken};
+
+use near_sdk::borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::Base58CryptoHash;
+use near_sdk::schemars::JsonSchema;
+use near_sdk::serde::{Deserialize, Serialize};
+use near_sdk::{AccountId, BlockHeight, Timestamp};
+
+pub type ProposalId = u64;
+
+type PostTag = String;
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(crate = "near_sdk::serde")]
+#[serde(tag = "proposal_version")]
+#[borsh(crate = "near_sdk::borsh")]
+#[schemars(crate = "near_sdk::schemars")]
+pub enum VersionedProposal {
+    V0(Proposal),
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(crate = "near_sdk::serde")]
+#[borsh(crate = "near_sdk::borsh")]
+#[schemars(crate = "near_sdk::schemars")]
+pub struct Proposal {
+    pub id: ProposalId,
+    pub author_id: AccountId,
+    #[serde(
+        serialize_with = "u64_dec_format::serialize",
+        deserialize_with = "u64_dec_format::deserialize"
+    )]
+    pub social_db_post_block_height: BlockHeight,
+    pub snapshot: ProposalSnapshot,
+    // // Excludes the current snapshot itself.
+    pub snapshot_history: Vec<ProposalSnapshot>,
+}
+
+impl From<VersionedProposal> for Proposal {
+    fn from(vp: VersionedProposal) -> Self {
+        match vp {
+            VersionedProposal::V0(v0) => v0,
+        }
+    }
+}
+
+impl From<Proposal> for VersionedProposal {
+    fn from(p: Proposal) -> Self {
+        VersionedProposal::V0(p)
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(crate = "near_sdk::serde")]
+#[borsh(crate = "near_sdk::borsh")]
+#[schemars(crate = "near_sdk::schemars")]
+pub struct ProposalSnapshot {
+    pub editor_id: AccountId,
+    #[serde(
+        serialize_with = "u64_dec_format::serialize",
+        deserialize_with = "u64_dec_format::deserialize"
+    )]
+    pub timestamp: Timestamp,
+    pub labels: HashSet<PostTag>,
+    #[serde(flatten)]
+    pub body: VersionedProposalBody,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(crate = "near_sdk::serde")]
+#[borsh(crate = "near_sdk::borsh")]
+#[schemars(crate = "near_sdk::schemars")]
+pub struct ProposalBodyV0 {
+    pub name: String,
+    pub category: String,
+    pub summary: String,
+    pub description: String,
+    pub linked_proposals: Vec<ProposalId>,
+    #[serde(
+        serialize_with = "u128_dec_format::serialize",
+        deserialize_with = "u128_dec_format::deserialize"
+    )]
+    pub requested_sponsorship_amount: Balance,
```

Comment:
Balance u128 is not going to play well with JSON serialization. Given that we only deal with integer amounts (without fractional part), let's just use u32 for the balance in proposals (try moving the Balance definition from lib.rs to post/mod.rs if that won't trigger changes outside of post/ module, so we can kill it together with the post module when the time will come).